src/Microsoft.DotNet.Interactive/KernelExtensions.cs

```diff
@@ -209,64 +211,90 @@ public static T UseDotNetVariableSharing<T>(this T kernel)
-            where TKernel : DotNetKernel
+            where TKernel : Kernel, ISupportGetValue
         {
-            kernel.AddDirective(who_and_whos());
+            kernel.AddDirective(who());
             Formatter.Register(new CurrentVariablesFormatter());
             return kernel;
         }
 
-        private static Command who_and_whos()
+        public static TKernel UseWhos<TKernel>(this TKernel kernel)
+            where TKernel : Kernel, ISupportGetValue
         {
-            var command = new Command("#!whos", "Display the names of the current top-level variables and their values.")
+            kernel.AddDirective(whos());
+            Formatter.Register(new CurrentVariablesFormatter());
+            return kernel;
+        }
+
+        private static Command who()
+        {
+            var command = new Command("#!who", "Display the names of the current top-level variables.")
             {
-                Handler = CommandHandler.Create((ParseResult parseResult, KernelInvocationContext context) =>
+                Handler = CommandHandler.Create(async (ParseResult parseResult, KernelInvocationContext context) =>
                 {
-                    var alias = parseResult.CommandResult.Token.Value;
-
-                    var detailed = alias == "#!whos";
+                    await DisplayValues(context, false);
+                })
+            };
 
-                    Display(context, detailed);
+            return command;
+        }
 
-                    return Task.CompletedTask;
+        private static Command whos()
+        {
+            var command = new Command("#!whos", "Display the names of the current top-level variables and their values.")
+            {
+                Handler = CommandHandler.Create(async (ParseResult parseResult, KernelInvocationContext context) =>
+                {
+                    await  DisplayValues(context, true);
                 })
             };
 
-            // TODO: (who_and_whos) this should be a separate command with separate help
-            command.AddAlias("#!who");
-
             return command;
+        }
 
-            void Display(KernelInvocationContext context, bool detailed)
+        private static async Task DisplayValues(KernelInvocationContext context, bool detailed)
+        {
+            if (context.Command is SubmitCode &&
+                context.HandlingKernel is ISupportGetValue)
             {
-                if (context.Command is SubmitCode &&
-                    context.HandlingKernel is DotNetKernel kernel)
+                var nameEvents = new List<ValueInfosProduced>();
+
+                var result = await context.HandlingKernel.SendAsync(new RequestValueInfos(context.Command.TargetKernelName));
+                result.KernelEvents.OfType<ValueInfosProduced>().Subscribe(e => nameEvents.Add(e));
+
+                var valueNames = nameEvents.SelectMany(e => e.ValueInfos.Select(d => d.Name)).Distinct().ToList();
```

Comment:
You don't need to be so aggressive with `ToList`.

Reply:
done